### PR TITLE
fix(cli/emit): Don't cache .tsbuildinfo unless emitting

### LINKED ISSUE #12755: `deno run` cannot find export in module that has export
The latest 1.16.1 has issues where it says that a module doesn't have an export when it clearly does. It is very difficult to reproduce this and so I haven't been able to create a simple example yet, but it's happened to me twice today during my development so I'm logging it (actually 3 times, but the third was slightly different and probably related).

The script is using this import:

```ts
import { createMinifier } from "https://deno.land/x/dts_minify@0.3.0/mod.ts";
```

Then see what occurred on my command line... the issue is fixed when I run it in an old version of deno (in this case 1.15):

```ts
> deno run -A scripts/createLibFile.ts
warning: Cannot load module "https://deno.land/x/dts_minify@0.3.0/mod.ts".
    at file:///V:/ts-morph/packages/common/scripts/createLibFile.ts:1:32
  If the source module contains only types, use `import type` and `export type` to import it instead.
error: Uncaught SyntaxError: The requested module 'https://deno.land/x/dts_minify@0.3.0/mod.ts' does not provide an export named 'createMinifier'
import { createMinifier } from "https://deno.land/x/dts_minify@0.3.0/mod.ts";
         ~~~~~~~~~~~~~~
    at <anonymous> (file:///V:/ts-morph/packages/common/scripts/createLibFile.ts:1:10)
> deno -V
deno 1.16.1
> bvm exec deno 1.15 run -A scripts/createLibFile.ts
Check file:///V:/ts-morph/packages/common/scripts/createLibFile.ts
> deno run -A scripts/createLibFile.ts
Check file:///V:/ts-morph/packages/common/scripts/createLibFile.ts
>
```

The issue can also be fixed by deleting the deno cache folder.

This might be related to #12740.

### LINKED ISSUE #12807: Skip loading TypeScript buildinfo when `--reload` flag is passed
Bundled this with `deno bundle --no-check mod.ts` and after that ran `deno run -A --unstable mod.ts` and now this happens. I'm not quite sure how but this happens with a dynamic/worker imports.

Code can be found on [MierenManz/cards](https://github.com/mierenmanz/cards)

```
thread 'worker-0' panicked at 'internal error: entered unreachable code: unexpected missing emit: file:///D:/programming/js_ts/card_game/utils/match/game_errors.ts', 
cli\proc_state.rs:544:15
```

```
stack backtrace:
   0:     0x7ff6174dc3a6 - <unknown>
   1:     0x7ff616c8cc3a - <unknown>
   2:     0x7ff6174da5f8 - <unknown>
   3:     0x7ff6174db462 - <unknown>
   4:     0x7ff6174daa7d - <unknown>
   5:     0x7ff616c419bf - <unknown>
   6:     0x7ff6174da427 - <unknown>
   7:     0x7ff6174e2b35 - <unknown>
   8:     0x7ff6174e2aa9 - <unknown>
  10:     0x7ff6185d5bc0 - CrashForExceptionInNonABICompliantCodeRange
  11:     0x7ff61697b3da - <unknown>
  12:     0x7ff61697888d - <unknown>
  13:     0x7ff616978566 - <unknown>
  14:     0x7ff616fe496e - <unknown>
  15:     0x7ff616fdfc46 - <unknown>
  16:     0x7ff616fdd912 - <unknown>
  17:     0x7ff616fdaa05 - <unknown>
  18:     0x7ff616fd9796 - <unknown>
  19:     0x7ff6174fdd2b - <unknown>
  20:     0x7fffcbcf7034 - BaseThreadInitThunk
  21:     0x7fffcc862651 - RtlUserThreadStart
  ```

### LINKED ISSUE #12832: panicked at 'internal error: entered unreachable code: unexpected missing emit
Running `deno run ./test.ts` creates this output:

>[0m[32mCheck[0m file:///Users/maxwellburson/workspace/benchmarkTest/test.ts
>thread 'main' panicked at 'internal error: entered unreachable code: unexpected missing emit: https://deno.land/x/fetchbase64@1.0.0/src/local.ts', cli/proc_state.rs:544:15
>stack backtrace:
>   0:        0x1051bccd8 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha47ac0a9192e239e
>   1:        0x104c4ab54 - core::fmt::write::hfbe6a7b828c32513
>   2:        0x1051b08a4 - std::io::Write::write_fmt::h34cd061a987d18fd
>   3:        0x1051b18d0 - std::panicking::default_hook::{{closure}}::h2b5c2a08263c3d54
>   4:        0x1051b1568 - std::panicking::default_hook::h0572aa020dd71058
>   5:        0x104c157c0 - deno::setup_exit_process_panic_hook::{{closure}}::he251a6b398b7f383
>   6:        0x1051b2268 - std::panicking::rust_panic_with_hook::h36a6e8d335334d30
>   7:        0x1051bd090 - std::panicking::begin_panic_handler::{{closure}}::h71decad72c0b1843
>   8:        0x1051bd010 - std::sys_common::backtrace::__rust_end_short_backtrace::hfd636421a82c4798
>   9:        0x1051b1d50 - _rust_begin_unwind
>  10:        0x10617f2e0 - core::panicking::panic_fmt::hebdff93de232aa26
>  11:        0x104a8ea68 - deno::proc_state::ProcState::prepare_module_load::{{closure}}::hdc42434c8ad65bc5
>  12:        0x104a8cefc - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h2a46327fed52214f
>  13:        0x104a8cd34 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h6b81a4cf63517ed8
>  14:        0x1049fc2c8 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h6fa4f6199afd8720
>  15:        0x1049fb41c - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1025300baad02af4
>  16:        0x104aae604 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h148a380254dc526c
>  17:        0x104aacab4 - deno::run_command::{{closure}}::h72484fc8c377f6dc
>  18:        0x104aabe44 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h45912926a885511f
>  19:        0x1049dfc7c - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h9418c5697b011832
>  20:        0x1049d5e40 - deno::main::h0532f6ec18a4c124
>  21:        0x1049d2c40 - std::sys_common::backtrace::__rust_begin_short_backtrace::h9a39dbb11315ea1b
>  22:        0x1049d2e88 - _main

## test.ts

> import Graph from 'https://deno.land/x/deno_chart/mod.ts';
> 
> const graph = new Graph({
>   titleText: 'Uptime',
>   xAxisText: 'Hours',
>   yAxisText: 'Day',
>   
>   yMax: 50,
>   bar_width: 25,
>   graphSegments_X: 18,
> 
>   xTextColor: 'rgba(255,255,255,1)',
>   xSegmentColor: 'rgba(255,255,255,0.5)',
>   yTextColor: 'rgba(255,255,255,1)',
>   ySegmentColor: 'rgba(255,255,255,0.5)',
> 
>   // Verbose Logging (Optional)
>   verbose: true,
> });
> 
> // Random Bar Generation with Colors!
> const COLORS = [
>   '#345C7D', '#F7B094', '#F5717F', '#F7B094',
>   '#6C5B7A',
> ]
> 
> for (let i = 0; i < 12; i++) {
>   const clr = COLORS[Math.floor(Math.random() * COLORS.length)];
>   const y = Math.floor(Math.random() * 50);
> 
>   graph.add({
>     val: i,
>     label: y.toString(),
>     color: clr,
>   });
> }
> 
> // Draw to Canvas Context & Save png image
> graph.draw();
> graph.save('image.png');
> 

## Machine info

deno 1.16.2

ProductName:    macOS
ProductVersion: 12.0.1
BuildVersion:   21A559

It's one of the new M1 macs

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fixes #12755.
Fixes #12807.
Fixes #12832.

### CODE REVIEW

File: cli/tests/integration/run_tests.rs [hunk 1]
```diff
@@ -2305,3 +2305,39 @@ fn issue12740() {
...
   assert!(!status.success());
 }
+
+/// Regression test for https://github.com/denoland/deno/issues/12807.
+#[test]
+fn issue12807() {
+  let mod_dir = TempDir::new().expect("tempdir fail");
+  let mod1_path = mod_dir.path().join("mod1.ts");
+  let mod2_path = mod_dir.path().join("mod2.ts");
+  let mut deno_cmd = util::deno_cmd();
```
Comment (bartlomieju) [inline cli/tests/integration/run_tests.rs:2315]: Nice test!

**Review by kitsonk — APPROVED** [review]
LGTM, thanks for addressing this!!!

(sorry for the delay, I was focused on trying to get something else done)

**Comment by ry** [comment]
This PR improved memory usage in the "hello" benchmark

Screen Shot 2021-12-01 at 8 35 50 AM

**Comment by nayeemrmn** [comment]
> but slowed down type checking in the "bundle" benchmark
 
@ry https://github.com/denoland/deno/issues/12881#issuecomment-977319401 I believe the type checking benchmark had been cheating until this PR, since individual runs don't appear to be given fresh DENO_DIRs and there is a `--reload` fix included.

**Pull request merged**